### PR TITLE
Fix Cooldown for Raise Dead (Ghoul)

### PR DESCRIPTION
**Pull request by ghost**
@joschiwald: Fix written on 31 Jan 2014.
@Nyeriah: suggested null pointer check.
@ariel-: supplied code for null pointer check.
@ccrs: supplied code (cat && spellInfo->CategoryRecoveryTime) to replace (cat && catRec)

Fixes issue #11538
Replaces #13941 and #13924

### CODE REVIEW

**Comment by tkrokli** [comment]
Labels: [Branch-3.3.5a] [Comp-Core] [Feedback-PatchFix]

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -22138,6 +22138,35 @@ void Player::SendCooldownEvent(SpellInfo const* spellInfo, uint32 itemId /*= 0*/
     data << uint32(spellInfo->Id);
     data << uint64(GetGUID());
     SendDirectMessage(&data);
+
+    uint32 cat = spellInfo->GetCategory();
+    uint32 catRec = spellInfo->CategoryRecoveryTime;
```
Comment (Runico) [inline src/server/game/Entities/Player/Player.cpp]: line 22143 is useless remove it 
uint32 catRec = spellInfo->CategoryRecoveryTime;

**Comment by DDuarte** [comment]
> Edit: @DDuarte : in #13948 you added the label [Sub-Spells], is that label in any way relevant here?

Well.. the pull request name references a spell and the word 'spell' is spelled 18 times in the code added. I think it is somewhat relevant.

**Comment by ccrs** [comment]
First of all, thanks for calling me stupid trollkli, second its not my foult if u dont do things right.
when i suggested to replace  "if (cat && catRec)" by "if (cat && spellInfo->CategoryRecoveryTime)"
its kinda OBVIOUS that the declaration of catRec should be erased just as @Runico said.

sigh....

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -26918,4 +26947,3 @@ void Player::SendSupercededSpell(uint32 oldSpell, uint32 newSpell)
     data << uint32(oldSpell) << uint32(newSpell);
     GetSession()->SendPacket(&data);
 }
-
```
Comment (Nyeriah) [inline src/server/game/Entities/Player/Player.cpp]: Uintended line removal?

**Comment by tkrokli** [comment]
> First of all, thanks for calling me stupid trollkli, second its not my fault if u dont do things right.

I did not intend to call you stupid. I called myself stupid for doing what I did.

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (tkrokli) [inline src/server/game/Entities/Player/Player.cpp]: I thought it was a forgotten unintended blank line. Sorry if it was wrong.

**Comment by tkrokli** [comment]
Sorry for bothering you with this, guys. I thought I was doing something constructive.

**Comment by DDuarte** [comment]
And you are doing something constructive. Way more than a lot of other people.

What's missing for this to be fixed? I can help you with that. Meet me at the IRC (irc.rizon.net @ #trinity)

**Comment by tkrokli** [comment]
Test done so far: created a bloodelf DK, went through the initial death knight quest line to get the character "out in the field" so to speak. When done in Orgrimmar, I went to Outland and started questing to find some humanoids to test on. Fair enough, Raise Dead (ghoul) works, both with and without the [Master of Ghouls] talent for "perma-ghoul" (pet instead of 1 minute duration). Each tooltip seems to be set updated correctly when the talents are changed both for Master of Ghouls and for Night of the Dead. In fact, while playing with this fix enabled, I have stopped thinking that this has been an issue at all, it feels like it never was wrong in the first place. :smiling_imp:

Anyway, except for testing with and without [Corpse Dust], with and without [Master of Ghouls], with and without [Night of the Dead], what else should I test to be able to say that I have tested it properly?

**Comment by DDuarte** [comment]
That is way more than enough tests, thanks tkrokli
I'll see if I can get this merged soon™

**Comment by ccrs** [comment]
I dont know exactly how the second cooldown of this spell is added to m_spellCooldowns, but i suppose that this is a better approach to the prblem (can it be done?):

1. search in spellinfo or somewhere else if the spell that is being handled has another cooldown or if it adds more than one cooldown when used.
2. if it's true, search in player cooldowns wether this cooldown was added or not.
3. if it's true, send the necessary packet.

if i just knew how to do point 1 :P

**Comment by tkrokli** [comment]
It is good that you keep trying to find out if it would be possible to make any improvements on this code. I just hope someone else will be able to answer your questions, since it is beyond my skills to answer them.

**Comment by Nyeriah** [comment]
Have been running with this code for years, other than the crash that we already addressed there are no other issues.

**Comment by Demonid** [comment]
@Nyeriah Was that crash fixed in the new code?

I'm using that code since the day @joschiwald posted it for the first time, no issues but the crash that @Nyeriah already mentioned.

**Comment by tkrokli** [comment]
Thank you for confirming this, both of you. :+1: No wonder the code worked so well for me. :smiling_imp:

**Comment by Nyeriah** [comment]
> @Nyeriah Was that crash fixed in the new code?
> 
> I'm using that code since the day @joschiwald posted it for the first time (One year or so), no issues but the crash that @Nyeriah already mentioned.

Yes, it was. It was just a null pointer crash.

**Comment by tkrokli** [comment]
> Yes, it was. It was just a null pointer crash.

OK, thanks, then I can safely leave this PR as it is until it can be merged.

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (tkrokli) [inline src/server/game/Entities/Player/Player.cpp]: Just found out why I made this change. When I clone the source from the TC branch, the file appears to have 2 blank lines at the end when I open Player.cpp in Notepad++ - hence the removal of a blank line at the end.

**Comment by tkrokli** [comment]
Looks like I was able to squash the commits, except for maybe adding all 3 headings into 1. I used GitExt Browse to pull and push, but used git bash to run the `git rebase -i HEAD~3` command.

**Comment by DDuarte** [comment]
GitExt -> Commit window -> Tick Amend Commit, change commit message, Commit & Push and then tick Force push in Show Options

https://www.atlassian.com/git/tutorials/rewriting-history/git-commit--amend

**Comment by tkrokli** [comment]
OK, thanks a lot for the patience and support. :+1: :smiley:

**Comment by tkrokli** [comment]
Unless there is something else you would like changed in this PR, I guess I should start browsing the issues again and see if there is something similar to be done with some other 3.3.5 issue. ;-)

**Comment by tkrokli** [comment]
Will any updates on my personal profile branch (changing the revision hash of this PR source) have any negative effects on when this PR can be merged? Do you prefer that I to leave my profile alone until it is merged, to avoid making more noise in the system?

**Comment by DDuarte** [comment]
Not sure what a profile branch is.
Any updates you do to your `3.3.5` branch will be reflected in this PR.
If you got others PR to do you need to create them in a different branch but the base branch cannot be your `3.3.5`, it must be a clean one.
That's why everywhere it is said that each PR should be in its own branch, so you can keep updating the main one without affecting the PR.

**Comment by tkrokli** [comment]
@soulfrost : are you 100% sure about this? If I test it on my build, will it remain just as stable as it is with the current code? (I'm just asking because I have very little faith in myself when updating a PR.)

**Comment by DDuarte** [comment]
The change that soulfrost pointed out does not change what the code does.

You don't need to update your `3.3.5` branch unless the PR cannot be merged any more by Github.

**Comment by tkrokli** [comment]
@DDuarte : Thanks for the confirmation. :+1:  ("If it works, don't fix it.") ;-)

**Comment by soulfrost** [comment]
@DDuarte just code style :)

**Comment by ccrs** [comment]
its the same code i pointed out <.<

**Comment by ccrs** [comment]
````c++
 uint32 cat = spellInfo->GetCategory();
    if (cat && spellInfo->CategoryRecoveryTime)
    {
        SpellCategoryStore::const_iterator ct = sSpellsByCategoryStore.find(CAT);
````
It'd have been a baaaaaaaad idea

**Comment by tkrokli** [comment]
@ccrs : Thanks, I suppose that means it is best to keep the code like it is, right? If I understand you correctly, you say that soulfrost's suggestion is not a good idea? :8ball: :suspect:

**Comment by ccrs** [comment]
exactly xD

**Comment by tkrokli** [comment]
`*sigh of relief*` Ah, thanks. I am relieved that I don't have to change the code, because it currently looks good in my server build running on my computer (base core : 4adf31c + tkrokli@073c14e, TDB 335.57 + updates => 4adf31c + manual additions: 0c9e39f + 03edeeb + https://github.com/TrinityCore/TrinityCore/issues/14020#issuecomment-72644028 (SQL query inserts)

**Comment by tkrokli** [comment]
Thanks a bunch, I will now merge the 3.3.5 source with my local branch. As for the 23 000 commits on the 3.3.5 branch - I suggest a "Ding!" sound just to mark the 23K milestone! :+1:

**Comment by declart** [comment]
it is still bugged i show you a pic.
In arena is very impossible to win without a ghoul..

http://i.imgur.com/2EgJ2po.jpg

https://github.com/TrinityCore/TrinityCore/commit/3e42d83e9cb86695a0467726f327599ff354293f

**Comment by tkrokli** [comment]
@declart : please post your issue separately or in any existing issue about Raise Dead.
This PR is merged and regarded as closed.

**Comment by declart** [comment]
ok

**Pull request merged**